crates/build/src/tests.rs

```diff
@@ -507,31 +510,97 @@ fn unchanged_contract_skips_optimization_and_metadata_steps(
+            dest_wasm_modified,
+            metadata_result_modified,
+            contract_bundle_modified,
+        )
+    }
+
     // when
     let res1 = super::execute(args.clone()).expect("build failed");
+    let (opt_result_modified1, metadata_modified1, contract_bundle_modified1) =
+        get_last_modified(&res1);
     let res2 = super::execute(args).expect("build failed");
+    let (opt_result_modified2, metadata_modified2, contract_bundle_modified2) =
+        get_last_modified(&res2);
 
     // then
-    assert!(
-        res1.optimization_result.is_some(),
-        "Initial build should perform wasm optimization"
+    assert_eq!(
+        opt_result_modified1, opt_result_modified2,
+        "Subsequent build of unchanged contract should not perform optimization"
     );
-    assert!(
-        res1.metadata_result.is_some(),
-        "Initial build should perform generate metadata"
+    assert_eq!(
+        metadata_modified1, metadata_modified2,
+        "Subsequent build of unchanged contract should not perform metadata generation"
     );
+    assert_eq!(contract_bundle_modified1, contract_bundle_modified2, "Subsequent build of unchanged contract should not perform contract bundle generation");
+
+    Ok(())
+}
+
+fn unchanged_contract_no_metadata_artifacts_generates_metadata(
+    manifest_path: &ManifestPath,
+) -> Result<()> {
+    let res1 = super::execute(ExecuteArgs {
+        manifest_path: manifest_path.clone(),
+        build_artifact: BuildArtifacts::CodeOnly,
+        ..Default::default()
+    })
+    .expect("build failed");
+
+    // CodeOnly should only generate Wasm code artifact
+    assert!(res1.dest_wasm.as_ref().unwrap().exists());
+    assert!(res1.metadata_result.is_none());
+
+    let dest_wasm_modified_pre = file_last_modified(&res1.dest_wasm.unwrap());
+
+    let res2 = super::execute(ExecuteArgs {
+        manifest_path: manifest_path.clone(),
+        build_artifact: BuildArtifacts::All,
+        ..Default::default()
+    })
+    .expect("build failed");
+
+    let dest_wasm_modified_post = file_last_modified(res2.dest_wasm.as_ref().unwrap());
+
+    // Code remains unchanged, but metadata artifacts are now generated
+    assert_eq!(dest_wasm_modified_pre, dest_wasm_modified_post);
     assert!(
-        res2.metadata_result.is_none(),
-        "Subsequent build should not perform wasm optimization"
+        res2.metadata_result
+            .as_ref()
+            .unwrap()
+            .dest_metadata
+            .exists(),
+        "Metadata file should have been generated"
     );
     assert!(
-        res2.metadata_result.is_none(),
-        "Subsequent build should not generate metadata"
+        res2.metadata_result.as_ref().unwrap().dest_bundle.exists(),
+        "Contract bundle should have been generated"
     );
 
     Ok(())
 }
 
+/// Get the last modified date of the given file.
+fn file_last_modified(path: &Path) -> SystemTime {
+    fs::metadata(path).unwrap().modified().unwrap()
```

Comment:
Should the `Result` be handled here? Are we sure that the file exists under given path?

Reply:
At the moment we are assuming the file exists, yes. In the cases it is used in the tests that should be true. That said I have replaced the `unwrap`s with `ok_or_else` to properly display the error message in case of an error to make it easier to debug these errors if they arise. https://github.com/paritytech/cargo-contract/pull/992/commits/dfc92152e9816aae151e5ae981411c293d5b8667